We can (now?) use value_to_string and avoid problems with custom fields.

lamby opened the pull request.

lamby committed:
We can (now?) use value_to_string and avoid problems with custom fields.

Signed-off-by: Chris Lamb <chris@chris-lamb.co.uk>

giftig commented:
Please limit the first line of your commit message to 50 characters maximum.
Please explain what sort of problems this commit avoids in the commit message.

danpalmer committed:
Merge pull request #1 from lamby/master

We can (now?) use value_to_string and avoid problems with custom fields.

danpalmer committed:
Support Django >= 1.6

danpalmer committed:
This wasn't supposed to be added

danpalmer committed:
parent_model was a private API and removed in Django 1.8

danpalmer committed:
Lookup the right model

danpalmer committed:
Special case some field types.

Otherwise we instantiate entities with a string representation

lamby committed:
This is now model_name

prophile committed:
This link is no longer live

prophile committed:
Correct these

prophile committed:
Add `install_requires`

prophile committed:
Exclude the "tests" directory

prophile committed:
Trailing comma

prophile committed:
Delete this documentation

It doesn't add any value here.

prophile committed:
Add a first test

prophile committed:
Add a simple (currently failing)_cached relation test

prophile committed:
Mark as skipped

prophile committed:
Add cache invalidation tests

prophile committed:
Fix this test

prophile committed:
Add a version number on the cache keys

prophile committed:
Run all encoding through Pickle

prophile committed:
Merge pull request #2 from thread/improvements

Use Pickle for encoding, which avoids the cache_toolbox encoding issues

prophile committed:
Fix the authentication middleware

At some point `request.session[SESSION_KEY]` started returning a `unicode`
rather than an `int`. As a result this would always fail, and thus always
fall back to the parent mechanism.

prophile committed:
Don't assume a numeric primary key

Commit:
Reflow

Commit:
Specify using for relations

This means that when there is a cache miss, the database that is currently in
use will be used to fetch the data.

For example, if you have a database slave, model Foo with a cached relation
bar, which has a property baz:
    random_foo = Foo.objects.order_by('?').using('slave').first()

    # On cache misses, this will hit the database slave, as that is what was
    # used to fetch baz.
    baz = random_foo.bar_cache.baz

lamby committed:
Releasing version 0.1.1

lamby committed:
Merge https://github.com/thread/django-cache-toolbox

* https://github.com/thread/django-cache-toolbox: (23 commits)
  Specify using for relations
  Reflow
  Don't assume a numeric primary key
  Fix the authentication middleware
  Run all encoding through Pickle
  Add a version number on the cache keys
  Fix this test
  Add cache invalidation tests
  Mark as skipped
  Add a simple (currently failing)_cached relation test
  Add a first test
  Delete this documentation
  Trailing comma
  Exclude the "tests" directory
  Add `install_requires`
  Correct these
  This link is no longer live
  Special case some field types.
  Lookup the right model
  parent_model was a private API and removed in Django 1.8
  ...

# Conflicts:
#	cache_toolbox/core.py
#	setup.py

lamby committed:
Releasing version 0.2.0

lamby committed:
Drop support for 1.6

lamby committed:
Releasing version 0.2.1

lamby committed:
This is now on related_model, not model!

lamby committed:
Releasing version 0.2.2

lamby committed:
Add missing trailing comma

lamby committed:
Drop broken creation non-UPSET functionality

Signed-off-by: Chris Lamb <chris@chris-lamb.co.uk>

lamby committed:
Fork

lamby committed:
Drop unused import

lamby committed:
Releasing version 0.2.3

lamby committed:
Move away from deprecated django.template.resolve_variable.

lamby committed:
Releasing version 0.2.4

giftig commented on COPYING:
```diff
@@ -1,3 +1,4 @@
+Copyright © 2016 Chris Lamb <chris@chris-lamb.co.uk>
```
Hahahahahahahahahahahahahahahaha.

PeterJCLaw committed:
Switch to the test running mechanism from django-enumfield

This works, whereas the previous 'python setup.py test' approach
didn't seem to work for me.

PeterJCLaw committed:
Support `hasattr`

We need to convert the `DoesNotExist` error into a
`RelatedObjectDoesNotExist` (which extends `DoesNotExist` by also
inheirting from `AttributeError`) so that `hasattr` can correctly
identify missing attributes.

PeterJCLaw committed:
Ensure `hasattr` is consistent

Previously accessing the natural (uncached) relation would cause
`hasattr` to always return True, even if the value wasn't actually
present. This changes that behaviour to be consistent with the
underlying descriptor by deferring to it in this case.

PeterJCLaw committed:
Ignore eggs (#1)

PeterJCLaw committed:
Django 1.x (#2)

* We're not compatible with Django 2.0 yet

* Add on_delete which is now expected

This will be required in Django 2.0.

lamby committed:
Releasing version 0.3.0

PeterJCLaw committed:
Avoid race condition between transaction commit and cache clear

See inline comment for details.

Co-Authored-By: Alistair Lynn <arplynn@gmail.com>

lamby committed:
Releasing version 0.3.1

PeterJCLaw committed:
Support older setuptools

It turns out that older setuptools requires that there be commas
in dependency specifiers while newer versions are a bit more relaxed.

Since older versions need to at least parse the specifiers in order
to upgrade themselves, we need to use the older format if we want
to support the older versions _at all_, even if we aren't targetting
them.

Obviously having made this change we don't actually need a newer
version of setuptools, so we don't add a specific dependency there.

Previously this would error with:

```
Collecting django-cache-toolbox==0.3.1
  Downloading https://files.pythonhosted.org/packages/eb/f7/f39a121ade50b873b8dbd9f285c3c02810d08e96404e1ca2d6546b48db86/django-cache-toolbox-0.3.1.tar.gz
    Complete output from command python setup.py egg_info:
    error in django-cache-toolbox setup command: 'install_requires' must be a string or list of strings containing valid project/version requirement specifiers

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-187om0y_/django-cache-toolbox/
```

lamby committed:
Releasing version 0.3.2

mthpower committed:
Upgrade for Django 2.0 compatibility

mthpower committed:
Update setup.py since the tests don’t pass under Django 1.8

lamby committed:
Releasing version 1.0.0

danpalmer committed:
Bump allowed Django version to <2.2

This appears to work fine in practice on 2.1.

lamby committed:
Releasing version 1.1.0

PeterJCLaw committed:
Add a README with basic usage information

PeterJCLaw committed:
Actually demonstrate using a _cached_ relation

The previous code worked, but wasn't actually showing what it
intended to show.

lamby committed:
Also support Django 2.2.

lamby committed:
Releasing version 1.1.1

PeterJCLaw committed:
Include the README as the package long description

PeterJCLaw committed:
Cache negative relation lookups locally

This matches what Django does.

Fixes https://github.com/lamby/django-cache-toolbox/issues/15

lamby committed:
Make setup.py executable.

lamby committed:
Releasing version 1.2.0

PeterJCLaw committed:
Fix a secondary error from cache exceptions

If cache.get manages to raise an unexpected exception, then
previously our relation handling would end up raising an
`UnboundLocalError` due to the use of `instance` in the
`finally` clause.

This changes that behaviour by duplicating the `setattr` call.

lamby committed:
Releasing version 1.2.1

PeterJCLaw committed:
Explicitly require cached relations are primary keys

This should prevent potential issues which could otherwise arise
from using the primary key of the current model in the cache key
for the cached model.

Fixes https://github.com/lamby/django-cache-toolbox/issues/22

lamby committed:
Move to Python 3.x.

lamby committed:
Releasing version 1.3.0

PeterJCLaw committed:
Make the README example more correct

PeterJCLaw committed:
Extract de/serialisation helpers

I'm about to make some changes in get_instance, so having these
separate will make it easier.

PeterJCLaw committed:
Extract a couple of utils

I'm about to need these in the core logic.

PeterJCLaw committed:
Support always fetching some relations when loading a model

This will be particularly useful when applied to relations on the
User model, which is loaded by our middleware. In that case,
this allows for efficient loading of several models upfront.

Fixes https://github.com/lamby/django-cache-toolbox/issues/21

matt-dalton committed:
Allow use of extended User model

lamby committed:
Releasing version 1.4.0

lamby committed:
Apply 'black'.

lamby committed:
Releasing version 1.5.0

PeterJCLaw committed:
Wrap these collections in tuples for compatibility

Django itself seems fine with these being dict_keys, however the
django-perf-rec library isn't. While it's their bug, fixing here
is easy enough.

PeterJCLaw committed:
Avoid unrelated select-relateds

It turns out that select_related() when called without arguments
(as might happen if we didn't have anything in our related_names
collection) will try to select related any relations which are
non-nullable. Since we don't want that, guard the addition of the
select_related to our queryset.

PeterJCLaw committed:
Cope with only some of the related models actually being loaded

The main thing this fixes is the case where the primary model is
not returned in the cached results. However it also changes the
handling in the case where one of the related models isn't found.
Previously the former would error and the latter would silently
assign `None` into the local relation cache field (and would thus
never be populated even if there should have been a value there).

lamby committed:
Releasing version 1.6.0

lamby committed:
Configure setup.cfg to generate a wheel during the release process.

lamby committed:
Releasing version 1.6.1
